rnd/autogpt_server/autogpt_server/executor/manager.py

```diff
@@ -350,76 +354,111 @@ def on_node_execution(cls, q: ExecutionQueue[NodeExecution], data: NodeExecution
     @classmethod
     def on_graph_executor_start(cls):
         cls.pool_size = Config().num_node_workers
-        cls.executor = ProcessPoolExecutor(
-            max_workers=cls.pool_size,
+        cls._init_node_executor_pool()
+        logger.warning(f"Graph executor started with max-{cls.pool_size} node workers.")
+
+    @classmethod
+    def _init_node_executor_pool(cls):
+        cls.executor = Pool(
+            processes=cls.pool_size,
             initializer=cls.on_node_executor_start,
         )
-        logger.warning(f"Graph executor started with max-{cls.pool_size} node workers.")
 
     @classmethod
-    def on_graph_execution(cls, graph_data: GraphExecution):
+    def on_graph_execution(cls, graph_data: GraphExecution, cancel: threading.Event):
         prefix = get_log_prefix(graph_data.graph_exec_id, "*")
         logger.warning(f"{prefix} Start graph execution")
 
+        finished = False
+
+        def cancel_handler():
+            while not cancel.is_set():
+                cancel.wait(1)
+            if finished:
+                return
+            cls.executor.terminate()
```

Comment:
To avoid dangling executor, can we put _init_node in finally block

Reply:
what do you mean?

Reply:
Can terminate() function fail and throws an error ? e.g: timeout, concurrent failure, etc.

In such cases the executor will be in the non-usable state, and the `cls._init_node_executor_pool()` will not be called making the executor faulty.

Reply:
If not feel free to resolve this :)